pkg/networkservice/acl/acl_parser.go

```diff
@@ -0,0 +1,273 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package acl
+
+import (
+	"context"
+	"strconv"
+	"strings"
+
+	"github.com/edwarnicke/govpp/binapi/acl_types"
+	"github.com/edwarnicke/govpp/binapi/ip_types"
+	"github.com/pkg/errors"
+
+	"github.com/networkservicemesh/sdk/pkg/tools/log"
+)
+
+const (
+	action        = "action"     // DENY, PERMIT, REFLECT
+	dstPrefix     = "dst_prefix" // IPv4 or IPv6 CIDR
+	srcPrefix     = "src_prefix" // IPv4 or IPv6 CIDR
+	protocol      = "proto"
+	icmpTypeFirst = "icmp_type_first" // 8-bit unsigned integer
+	icmpTypeLast  = "icmp_type_last"  // 8-bit unsigned integer
+	icmpCodeFirst = "icmp_code_first" // 8-bit unsigned integer
+	icmpCodeLast  = "icmp_code_last"  // 8-bit unsigned integer
+	srcPortLast   = "src_port_last"   // 16-bit unsigned integer
+	srcPortFirst  = "src_port_first"  // 16-bit unsigned integer
+	dstPortLast   = "dst_port_last"   // 16-bit unsigned integer
+	dstPortFirst  = "dst_port_first"  // 16-bit unsigned integer
+	tcpFlagsMask  = "tcp_flags_mask"  // 8-bit unsigned integer
+	tcpFlagsValue = "tcp_flags_value" // 8-bit unsigned integer
+)
+
+func parseACLRulesMap(ctx context.Context, rules map[string]string) (aclRules []acl_types.ACLRule) {
+	logger := log.FromContext(ctx).WithField("acl", "parser")
+	for _, r := range rules {
+		rule := &acl_types.ACLRule{}
+
+		parsed := parseKVStringToMap(r, ",", "=")
+
+		if err := addAction(parsed, rule); err != nil {
+			logger.Errorf("Error parsing action: %v", err.Error())
+			continue
+		}
+
+		if err := addProtocol(parsed, rule); err != nil {
+			logger.WithField("acl", "parser").Errorf("Error parsing protocol: %v", err.Error())
+			continue
+		}
+
+		addPrefixes(parsed, rule)
+
+		if err := addSrcPortOrIcmpType(parsed, rule); err != nil {
+			logger.WithField("acl", "parser").Errorf("Error parsing srcPortOrIcmpType: %v", err.Error())
+			continue
+		}
+
+		if err := addDstPortOrIcmpCode(parsed, rule); err != nil {
+			logger.WithField("acl", "parser").Errorf("Error parsing dstPortOrIcmpType: %v", err.Error())
+			continue
+		}
+
+		addTCPFlags(parsed, rule)
+
+		aclRules = append(aclRules, *rule)
+	}
+
+	return aclRules
+}
+
+var actionMap = map[string]acl_types.ACLAction{
+	"permit":  acl_types.ACL_ACTION_API_PERMIT,
+	"reflect": acl_types.ACL_ACTION_API_PERMIT_REFLECT,
+	"deny":    acl_types.ACL_ACTION_API_DENY,
+}
+
+func addAction(parsed map[string]string, rule *acl_types.ACLRule) error {
+	actionName, ok := parsed[action]
+	if !ok {
+		return errors.New("action is missing")
+	}
+
+	a, ok := actionMap[actionName]
+	if !ok {
+		return errors.New("invalid action")
+	}
+
+	rule.IsPermit = a
+	return nil
+}
+
+func addPrefixes(parsed map[string]string, rule *acl_types.ACLRule) {
+	dst, dstOk := parsed[dstPrefix]
+	if !dstOk {
+		rule.DstPrefix = ip_types.Prefix{}
+	} else {
+		dstpref, err := ip_types.ParsePrefix(dst)
+		if err != nil {
+			rule.DstPrefix = ip_types.Prefix{}
```

Comment:
Probably it would be better to return an error in such case?

Reply:
done